07_errors/jimbotech/storer_test.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+)
+
+type storesSuite struct {
+	suite.Suite
+	store  Storer
+	mapper mapTest
+}
+
+const brown = "brown"
+const black = "black"
+const grey = "grey"
+
+func TestSuite(t *testing.T) {
+	suite.Run(t, &storesSuite{store: NewMapStore()})
+	suite.Run(t, &storesSuite{store: &SyncMapStore{}})
+}
+
+//SetupTest creates the correct empty map for each test
+func (s *storesSuite) SetupTest() {
+	switch s.store.(type) {
+	case MapStore:
+		s.store = NewMapStore()
+	case *SyncMapStore:
+		s.store = &SyncMapStore{}
+	default:
+		s.Fail("Unknown Storer implementation")
+	}
+	s.mapper = s.store.(mapTest)
+}
+
+func (s *storesSuite) TestReadSuccess() {
+	pup := create(s)
+	// now check by reading the value back and compare
+	pup2, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal(brown, pup2.Colour)
+	// modify the retured value to make sure the
+	// value in the store does not change
+	pup2.Colour = grey
+	pup3, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal(brown, pup3.Colour)
+	s.NotEqual(pup2, pup3)
+}
+
+// TestCreateSuccess add to the store and verify
+// by reading that it is in the store
+func (s *storesSuite) TestCreateSuccess() {
+	pup := create(s)
+	// Now modify the original and make sure the
+	// value in the store will not change
+	pup.Colour = black
+	// now check by reading the value back and compare
+	pup2, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal("kelpie", pup2.Breed)
+	s.Equal(brown, pup2.Colour)
+	s.Equal("indispensable", pup2.Value)
+	s.True(pup2.Colour == brown)
+	s.True(pup.Colour == black)
+	s.NotEqual(pup, pup2)
+}
+
+func create(s *storesSuite) *Puppy {
+	pup := Puppy{Breed: "kelpie", Colour: brown, Value: "indispensable"}
+	id, err := s.store.CreatePuppy(&pup)
+	s.Require().NoError(err)
+	s.Require().NotEqual(pup.ID, uint32(1))
+	s.Require().Equal(id, pup.ID, "Pup id must be set to actual id")
```

Comment:
This require/assert belongs in `TestCreateSuccess` as that is an invariant of the `CreatePuppy` method. This function is a text fixture called from many places, which means if the storer returned a bad id such that this fails, every test that uses this fixture function will also fail even though what they are testing themselves may well be correct.

Reply:
Moved